File: app/src/main/java/com/suzei/racoon/ui/friend/FriendPresenter.java
package com.suzei.racoon.ui.friend;

import android.widget.Toast;

import com.google.firebase.database.DatabaseError;
import com.suzei.racoon.ui.base.Contract;
import com.suzei.racoon.ui.friend.FriendInteractor;

public class FriendPresenter implements Contract.Listener<String> {

    private Contract.DetailsView<String> friendView;
    private FriendInteractor friendInteractor;

    public FriendPresenter(Contract.DetailsView<String> friendView) {
        this.friendView = friendView;
        friendInteractor = new FriendInteractor(this);
    }

    public void sendRequest(String currentUserId, String friendId, String currentState) {
        friendInteractor.executeRequest(currentUserId, friendId, currentState);
    }

    public void readCurrentState(String currentUserId, String friendId) {
        friendInteractor.readCurrentState(currentUserId, friendId);
    }

    public void destroy(String currentUserId, String friendId) {
        friendInteractor.destroyReading(currentUserId, friendId);
    }

    @Override
    public void onLoadSuccess(String data) {
        friendView.onLoadSuccess(data);
    }

    @Override
    public void onLoadFailed(String message) {
        friendView.hideProgress();
        friendView.onLoadSuccess(message);
    }
}

File: app/src/main/java/com/suzei/racoon/ui/friend/FriendPresenter.kt
package com.suzei.racoon.ui.friend

import com.suzei.racoon.model.Users

class FriendPresenter(
    private val mUser: Users?,
    private val mView: FriendContract.View
) : FriendContract.Presenter {
    init {
        mView.setPresenter(this)
    }

    private val friendInteractor: FriendInteractor = FriendInteractor(this)

    override fun sendRequest(currentUserId: String?, friendId: String?, currentState: String?) {
        friendInteractor.executeRequest(currentUserId, friendId, currentState)
    }

    override fun readCurrentState(currentUserId: String?, friendId: String?) {
        friendInteractor.readCurrentState(currentUserId, friendId)
    }

    override fun destroy(currentUserId: String?, friendId: String?) {
        friendInteractor.destroyReading(currentUserId, friendId)
    }

    override fun setupDetails() {
        mUser?.let {
            mView.showDetails(it)
        }
    }

    override fun onLoadSuccess(data: String) {
        mView.onLoadSuccess(data)
    }

    override fun onLoadFailed(message: String) {
        mView.hideProgress()
        mView.onLoadSuccess(message)
    }
}